New: Add no-await-in-loop rule

nmote opened the pull request:
**What is the purpose of this pull request? (put an "X" next to item)**

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[x] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

**Please describe what the rule should do:**

This rule disallows the use of `await` within the body of a loop.

**What category of rule is this? (place an "X" next to just one item)**

[ ] Enforces code style
[x] Warns about a potential error
[ ] Suggests an alternate way of doing something
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
async function foo(things) {
  const results = [];
  for (const thing of things) {
    results.push(await bar(thing));
  }
  return baz(results);
}

async function baz() {
  for (var i; await bar(i); i++) {
  }
}
```

**Why should this rule be included in ESLint (instead of a plugin)?**

Using `await` inside of a loop is a bad pattern no matter what codebase you are operating in. There are occasional places where it is desirable but it is typically a sign that you have done something wrong. As ES2017 features become more widely adopted it is important that the tools help people avoid common pitfalls.

**What changes did you make? (Give an overview)**

I added the `no-await-in-loop` rule, along with tests and docs.

Fixes #4166

**Is there anything you'd like reviewers to focus on?**

* I used a `Set` which I assume is okay now because I saw it used in the `class-methods-use-this` rule.
* I didn't come across a precedent for my `wrapTest` function but I think it is worthwhile to not have to repeat the parser config in each test case.

platinumazure commented on lib/rules/no-await-in-loop.js:
```diff
@@ -0,0 +1,75 @@
+                        ) {
+                            context.report(
+                                node,
+                                `Avoid using await inside a loop. Consider refactoring to use Promise.all. If you are sure you want to do this, add \`// eslint-disable-line "}${context.id}\` at the end of this line.`
```
This is a pretty long message, and the bit about disabling a rule is universal to all rules. I would suggest removing that bit.

nzakas replied:
Agree. Our general format is "Unexpected x", so something like "Unexpected 'await' inside a loop." would fit here. The docs are there for people to read up if they don't know how to fix it.

platinumazure commented on lib/rules/no-await-in-loop.js:
```diff
@@ -0,0 +1,75 @@
+"use strict";
+
+// Node types which are considered loops.
+const loopTypes = {
```
This could be a Set.

nzakas replied:
:+1: Let's use sets wherever it makes sense.

platinumazure commented on lib/rules/no-await-in-loop.js:
```diff
@@ -0,0 +1,75 @@
+
+// Node types at which we should stop looking for loops. For example, it is fine to declare an async
+// function within a loop, and use await inside of that.
+const boundaryTypes = {
```
This could be a Set.

nmote commented:
I'll happily move to sets in the places you suggested. I was converting this rule from an environment that didn't have sets.

There's some context about disabling the rule in #4166. I think that because this is more of a suggestion than hard-and-fast rule. It's usually right, but there are legitimate use-cases for using `await` in a loop. Therefore I think that in this rule, it is more useful to explain how to disable it than in most rules. @kaicataldo do you have more thoughts on this? I'll remove it if there's consensus but I think it's worth considering.

not-an-aardvark commented:
I would agree with @platinumazure that we should remove the disabling instructions from the rule message. I think it's redundant given that the disabling syntax would apply to all rules. Maybe an ESLint formatter would want to display instructions for disabling rules, but I think it's out of place for a report message.

nmote commented:
Cool, I'll just wait to hear from @kaicataldo and assuming he agrees I'll make the change. Thanks for the feedback.

kaicataldo commented:
> I think that because this is more of a suggestion than hard-and-fast rule. It's usually right, but there are legitimate use-cases for using await in a loop. Therefore I think that in this rule, it is more useful to explain how to disable it than in most rules. 

I actually agree with this, and this is part of why I'm not a fan of relying on the standard eslint-disable comments. It *IS* weird to reference the non-rule-specific disable method in an error message, but I also think it's important for users to understand that there might be some legitimate use cases for this.

My preference is still to have a rule-specific way of marking an occurrence as okay.

nmote commented:
Okay, suggestions for such a mechanism? @platinumazure, @not-an-aardvark: what do you think about this idea?

not-an-aardvark commented:
Maybe we could support comment like `// falls through` in [`no-fallthrough`](http://eslint.org/docs/rules/no-fallthrough)? (We would have to figure out a brief way to describe the issue in a comment, though.)

My preference is still to have people use `eslint-disable` comments but not to include such a recommendation in the message. But I'm open to adding a specific opt-out comment for this rule.

nmote commented:
Ideas for phrasing? `// allow await` ?

kaicataldo commented:
I like `//allow await`. Succinct and clear. I'm not sure on whether the comment should be required on the line before each `await` expression or if it should be a block-level thing (disabling it for the the entire block of a loop). I'm leaning towards it being required for each `await` expression (even though it's potentially more verbose, it forces each use of `await` to be a conscious decision, which I think is the spirit of this rule.

nmote commented:
Agreed. It's more obvious when comments like this have local effects. I think the implementation will also be easier since I can copy almost directly from `no-fallthrough`.

not-an-aardvark commented:
I slightly disagree about allowing it for every `await` expression; that seems a bit tedious. I think having one comment per loop would be sufficient, and it would still be clearly a conscious decision:

```js
async function foo() {
  for (const bar of baz) {
    // allow await [in this loop]

    await foo;
    await bar;
  }
}
```

kaicataldo commented:
Yeah, I'm not sure which I prefer. If someone is modifying a large pre-existing loop, it could be missed. But agreed, it could be tedious. We just have to way that against how explicit we want it to be.

not-an-aardvark commented:
I think if someone is modifying a large pre-existing loop, it would only be dangerous if they didn't notice the comment and tried to refactor it into `Promise.all`. But that would require them to remove the entire loop, at which point they would almost certainly notice the comment even if the loop was large.

nzakas commented on docs/rules/no-await-in-loop.md:
```diff
@@ -0,0 +1,61 @@
+`await` as part of each operation is an indication that the program is not taking full advantage of
+the parallelization benefits of `async`/`await`.
+
+Usually, the code should be refactored to create all the promises at once, then get access to the
```
Please so show code examples in this section.

nzakas commented on lib/rules/no-await-in-loop.js:
```diff
@@ -0,0 +1,75 @@
+module.exports = {
+    meta: {
+        docs: {
+            description: "disallow accidental serial execution of parallelizable operations",
```
Maybe something clearer like "disallow await inside of loops"?

kaicataldo commented:
As I said in the issue, I don't want to hold this up over the content of the disable content, as that seems like a minor detail. If everyone else is fine with using the built-in `eslint-disable` comments, I'll retract my objection.

nmote commented:
Great, looks like we have consensus. Tomorrow I'll reword the message and address the other comments.

not-an-aardvark commented on tests/lib/rules/no-await-in-loop.js:
```diff
@@ -0,0 +1,87 @@
+ * @param {?string} error: The expected error, if any
+ * @returns {Object}: The object that describes the test case to the RuleTester
+ */
+function wrapTest(code, error) {
```
This function is fine with me, but one alternative would be to put the defaults in the `RuleTester` constructor:

```js
const ruleTester = new RuleTester({ parserOptions: { ecmaVersion: 2017 } });
```

nmote replied:
Oh cool, I didn't know you could do this

not-an-aardvark commented on tests/lib/rules/no-await-in-loop.js:
```diff
@@ -0,0 +1,87 @@
+        // Blocked by a class method,
+        wrapTest("async function foo() { while (true) { class Foo { async foo() { await bar; } } } }"),
+
+    ],
```
There are a few additional tests that I think would be nice to have:

```js
async function foo() {

  for (var bar of await baz) {} // should not report an error
  for (var bar in await baz) {} // should not report an error

  // make sure that it works without block loop bodies
  for (var bar in baz) await qux; // should report an error

  // (legacy syntax, only allowed in non-strict mode)
  for (var bar = await baz in qux) {} // should not report an error
}
```

nmote replied:
Sure, will add.

nmote commented:
Ping -- this should be good to go and easy to stamp at this point, right?

vitorbal commented on docs/rules/no-await-in-loop.md:
```diff
@@ -0,0 +1,89 @@
+# Disallow accidental serial execution of parallelizable operations (no-await-in-loop)
```
Could you also update this to reflect the new `meta.description` field of the rule?

nzakas replied:
I'd rather update this to say "Disallow await in loops" to match `meta.description`. The simpler explanation is much clearer.

vitorbal commented:
This looks good to me, except for one thing in the docs. Thanks, @nmote!

nmote commented:
@vitorbal will you merge it once I make those changes or do we have to wait for everyone who reviewed to accept?

platinumazure commented on lib/rules/no-await-in-loop.js:
```diff
@@ -0,0 +1,75 @@
+                            ancestorSet.has(ancestor.test) ||
+                            ancestorSet.has(ancestor.update)
+                        ) {
+                            context.report(
```
Could you please use the new-style `context.report()` call? This takes an object and looks something like this:

```js
context.report({
    node,
    message: "Unexpected `await` `inside a loop."
});
```

nmote commented:
@platinumazure Will you merge once I've made these two changes? Otherwise I will wait for everyone who needs to approve to request their changes so I can lump them in together as a single work item.

vitorbal commented:
@nmote we can only merge once all the reviewers that have requested changes do another review to approve the changes.

platinumazure commented:
@nmote Please feel free to ping reviewers when you have addressed their changes-- as I noted earlier, GitHub doesn't have a notification mechanism, and unfortunately we're all pretty swamped here. Thanks so much for your patience and I really look forward to seeing this get in!

nmote commented:
Great, thank you -- please excuse my ignorance about how the GitHub review mechanism works.

@nzakas have I addressed your concerns? Once you respond I'll go ahead with the changes that @vitorbal and @platinumazure requested.

platinumazure commented:
Thanks @nmote. If you want, you can incorporate @vitorbal's and my feedback and push, and then we can approve if all looks good, leaving only nzakas. But if you'd rather do it all at once, that's okay too.

We can squash your commits on merge, so feel free to dump as many commits into your branch as you need :smile:

nzakas commented on docs/rules/no-await-in-loop.md:
````diff
@@ -0,0 +1,89 @@
+asynchronous operations that were unsuccessful. In such cases it makes sense to use `await` within a
+loop and it is recommended to disable the rule via a standard ESLint disable comment:
+
+```js
````
These examples are confusing because they don't show actual code. I think you can safely remove everything after this point, or if you feel like a code sample is important, then please use a complete code example.

kaicataldo commented:
@nmote Friendly ping. Wanted to just make sure everything is clear -- let us know if something isn't!

nmote commented:
Yep, just waiting until I have a chance to circle back to this. Been pretty busy this week.

kaicataldo commented:
@nmote Totally understand - just wanted to make sure everything was clear. Thanks for working on this!

nmote commented:
I think I've addressed everything. @nzakas sorry to hear you haven't been feeling well, I hope things improve. My mother has chronic Lyme disease so I've seen firsthand how debilitating it can be.

CC @platinumazure @vitorbal -- does this look good?

kaicataldo commented on docs/rules/no-await-in-loop.md:
```diff
@@ -0,0 +1,78 @@
+
+This rule disallows the use of `await` within loop bodies.
+
+## Options
```
I think it might actually be clearer if we simply omit this section

nmote replied:
Sure

kaicataldo commented on docs/rules/no-await-in-loop.md:
````diff
@@ -0,0 +1,78 @@
+}
+```
+
+## When Not To Use It
````
Could we add a code example of disabling the comment here as well?

kaicataldo replied:
Fair enough. Thanks for the explanation :+1:

kaicataldo commented on tests/lib/rules/no-await-in-loop.js:
```diff
@@ -0,0 +1,69 @@
+        "async function foo() { while (true) { var y = async () => await foo; } }",
+        "async function foo() { while (true) { var y = async () => { await foo; } } }",
+
+        // Blocked by a class method,
```
Nit: Extra comma

kaicataldo commented:
Merged, as it looks as though @vitorbal and @nzaka's concerns were addressed. Thanks for contributing to ESLint!

jakearchibald commented:
I'm a little concerned about this one. I agree that going sequential is a bad thing, but waiting for `Promise.all` when you could start acting on data much sooner is also bad. Eg https://developers.google.com/web/fundamentals/getting-started/primers/async-functions#example_outputting_fetches_in_order

Feels like this rule should be opt-in or come with much heavier caveats.

jakearchibald commented:
Also, given this stance, will eslint warn on all [async iterators](https://github.com/tc39/proposal-async-iteration)?

not-an-aardvark commented:
@jakearchibald Your concerns are well noted. We tried to make it clear in the [`rule docs`](http://eslint.org/docs/rules/no-await-in-loop#when-not-to-use-it) that there are a lot of valid use cases for disabling this rule (and as you noted above, the rule is opt-in).

We haven't discussed what to do about async iterators yet, since ESLint generally doesn't support features until they reach stage 4. That said, I would be surprised if we decided to make the rule warn on them, because that would defeat the point of using async iterators at all.

nmote commented:
@jakearchibald Your concerns are why this rule is opt-in. As you point out there are many valid reasons to use `await` inside of a loop. However enabling this rule will make sure that people stop and think before doing so. In my mind, adding an occasional disable comment is a small price to pay to make sure that people think this through.

You're also right that it's more nuanced than just adding `Promise.all`. Because of that I would be open to rewording the message, but do note that one of the concerns on this PR was that the message was too long.

jakearchibald commented:
I'm not sure what the answer is, but presently this rule warns if you do it the worst way, *and* if you do it the best way.
